analytics/certificates-jan2025.md

```diff
@@ -0,0 +1,115 @@
+
+The non-persistent pools are subject to sortition for each vote, based on an updated stake distribution where the persistent voters have been removed and where the distribution is normalized to unit probability. The VRF value for that sortition is the bytes of the BLS signature on the election identifier $eid$. The probability that a pool with fraction $\sigma$ of the stake is awarded $k$ votes of the committee of $n$ votes is 
+
+$$
+\mathcal{P}(k) := \frac{(n \cdot \sigma)^k \cdot e^{- n \cdot \sigma}}{k!}
+$$
+
+This VRF value is used to look up the correct number of votes from the cumulative distribution for $\mathcal{P}(k)$. The same Taylor-series expansion technique used in Praos can handle the irrational arithmetic in a deterministic manner. In practice it is unlikely that the non-persistent pools would ever be awarded more than one vote, so it may be feasible to simply award one vote whenever $k \ge 1$: in such case the Peras voting sortition could use code very similar to the existing Praos block sortition.
+
+Each vote has a weight, with weights summing to unity. A quorum is achieved if the weights total to 0.75 or more.
+
+
+### Certification
+
+Consider the committee size $n$, which contains $m$ persistent voters.
+
+The certificate must contain the following information:
+
+- Common to all votes
+    - Election ID: Presumably a 32-byte identifier for the Peras election is included in the certificate, though perhaps this is not strictly necessary. This could just be the slot number.
+    - Message: the 32-byte hash of the Praos block being boosted is also included in the certificate.
+- Identity of voters
+    - Persistent voters are encoded in a bitset of size $m$, occupying $\left\lceil m / 8 \right\rceil$ bytes.
+    - Non-persistent voters are encoded by their Pool ID (or equivalent), occupying 32 bytes each and hence $32 \cdot (n - m)$ bytes total.
+    - Alternatively, all possible voters could be assigned bits in a bitset, with size $\left\lceil N_\text{pools} / 8 \right\rceil$.
+- Eligibility proof
+    - Persistent voters are eligible by definition (by virtue of their stake in the epoch), so no proof is needed.
+    - Non-persistent voters prove eligibility with a 96-byte BLS signature on the message, occupying $96 \cdot (n - m)$ bytes total.
+- Aggregate signatures
+    - Signed message: This aggregate BLS signature on the message is 96 bytes.
+    - Signed election proofs: Perhaps not strictly necessary, but another 96 byte BLS signature can attest to the proof of the eligibility, see **BLS.BSig** in [the Leios paper](https://iohk.io/en/research/library/papers/high-throughput-blockchain-consensus-under-realistic-network-assumptions/).
+    
+Thus the total certificate size is
```

Comment:
Similarly, is the following for the vote size correct (ignoring CBOR overhead again)?

 - Peras round number (8 bytes)
 - Pool ID (32 bytes)
 - Block hash that is being voted for (32 bytes)
 - Signature (92 bytes)
 - If non-persistent: eligibility proof (92 bytes)

So 164 bytes for votes from persistent voters, and 256 bytes for votes from non-persistent voters (or maybe a bit more for the latter if non-persistent voters can be elected multiple times with non-negligible probability, in which case it probably makes sense to record that weight in the vote).

If so could we record that here?

Reply:
One point that I mentionned several times is that, while the eligibility proof could be removed with Fait accompli, we may still want to include a proof of the stake of the voter (which includes their key and stake).

This touches on the [chain of (stake) certificate that Mithril](https://mithril.network/doc/mithril/advanced/mithril-protocol/certificates/) is working on.

Reply:
Can you expand on this/where can I read more about this?

I don't immediately see why you would need that:

 - Votes include the pool ID, so we can just look up their stake in the stake distribution, I don't see a way to fake anything here.
 - Certificates contain a bitset indicating who the persistent voters are, so we can look up their stake in the stake distribution, and (together with the signature verifiation), we check whether the certificate indeed reaches a quorum.

Reply:
I agree with @amesgen that the stake distribution is already attested to each epoch. So long as we base sortition on the stake at the epoch boundary, we don't need a separate proof.

Also, in 1bc140b560ca66522ccb6f4124f01df39dcb34ff, I revised several of the sizes downward:

- It turns out that the Pool ID is 28 bytes instead of 32 bytes.
- Yes, we can use the 8 byte slot number for the election ID.
- We can use BLS compression
    - Public keys are 96 bytes
    - Signatures are 48 bytes

Reply:
Thanks, updated the size of votes above based on that :+1:

Reply:
This would potentially be needed if we want to SNARK the certificate: we would need to ensure that the right stake was used when generating the votes for one certificate, hence we would need the stake distribution to be committed at some point in time.